### ISSUE TITLE
Inconsistent step values when using multiple val/test dataloaders when logged on step-level

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#11616]
## 🐛 Bug

Currently, the `LoggerConnector` handles the val_step index for step-level logging for val/test steps, due to which it shows incorrect step (x-axis) in case of multiple dataloaders.

### To Reproduce

<details><summary>CODE</summary>
<p>

```py
import torch
from torch.utils.data import DataLoader, Dataset
import time

from pytorch_lightning import LightningDataModule, LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size: int, length: int):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        return {"loss": loss}

    def validation_step(self, batch, batch_idx, dataloader_idx=None):
        loss = self(batch).sum()
        self.log('valid_loss', loss, on_step=True)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        return optimizer

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64))

    def val_dataloader(self):
        return [DataLoader(RandomDataset(32, 64)), DataLoader(RandomDataset(32, 64))]
```
</p>
</details>

```py
model = BoringModel()
trainer = Trainer(max_epochs=1)
trainer.fit(model)
```

Output:
The log for dataloader_idx=1 starts from 64, which is the step right after the first dataloader is completed.

<img width="821" alt="Screenshot 2022-01-25 at 5 10 43 PM" src="https://user-images.githubusercontent.com/30778939/150988699-8c3e02e0-6103-40b0-bb10-f386b9756012.png">

Expected:
for each datalaoder, it should start from 0.

If we go a little more crazy and do:

```py
model = BoringModel()
trainer = Trainer(max_epochs=1)
trainer.fit(model)
trainer.validate(model)
```

Output:
![Screenshot 2022-01-25 at 7 17 09 PM](https://user-images.githubusercontent.com/30778939/150989102-f1025e8e-927b-47f2-8100-0042a7574cbc.png)

Expected:
I think it should replace the logs created during fit in such a case.

### Proposal
If we consider the above issues:
I think a better solution would be to let `EvaluationLoop` track the logging step for each dataloader, instead of logger connector, in a list where each index represents the current dataloader and will be the only one that should be incremented.

But there can be another issue, where the user resets the dataloader after every n epochs, then in such a case, should we reset the list if we go with this? The reason is, we can't for sure say whether the new dataloaders will be the same/more/less, and if we don't reset, it might be a problem.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): master
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.0
- OS (e.g., Linux): mac
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

Open to better alternative solutions.

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @justusschock @ninginthecloud

**Comment by andreimargeloiu** [comment]
When is the bug planned to be fixed?

**Comment by rohitgr7** [comment]
soon!
